### PR TITLE
Support forcing modchat on ladder tour battles

### PR DESCRIPTION
**Pull request by mia-pi-git**
Requested here https://www.smogon.com/forums/threads/allow-for-auto-modchatting-on-a-prefix.3682737/#post-8829088.
Monkeypatches are here https://gist.github.com/mia-pi-git/61fc91671107113a252d9d39791bad87.
I was told it would be optimal to have this live on the server by the 25th, as there's a tour that'd benefit from this.

### CODE REVIEW

File: server/users.ts [hunk 1]
```diff
@@ -1480,9 +1480,13 @@ export class User extends Chat.MessageContext {
 			this.registered ? `[registered]` :
 			``;
 	}
-	battlesForcedPublic() {
-		if (!Config.forcedpublicprefixes) return null;
-		for (const prefix of Config.forcedpublicprefixes) {
+	forcedPublic(key: 'modchat' | 'privacy') {
```
Comment (Zarel) [inline server/users.ts:1483]: This should probably be `battleForcedSetting`?

Comment (Zarel) [reply server/users.ts:1483]: Also, putting it in `room-battle` might make maintenance easier than having the functionality spread out over lots of different files.

File: server/rooms.ts [hunk 2]
```diff
@@ -1935,13 +1933,19 @@ export const Rooms = {
 			} else if (playerOptions && playerOptions.hidden) {
 				privacySetter.add(playerOptions.user.id);
 			}
+			if (playerOptions) battle.checkForcedSettings(playerOptions.user);
+		}
+
+		// Special battles have modchat set to Player from the beginning
+		if (p1Special || battle.forcedSettings.modchat) {
+			room.settings.modchat = '\u2606';
```
Comment (Zarel) [inline server/rooms.ts:1941]: This can also be moved inside `checkForcedSettings`.

File: server/rooms.ts [hunk 3]
```diff
@@ -1935,13 +1933,19 @@ export const Rooms = {
...
 		}
 
 		if (privacySetter.size) {
-			if (battle.forcePublic) {
+			if (battle.forcedSettings.privacy) {
```
Comment (Zarel) [inline server/rooms.ts:1945]: Honestly, this all can.

File: server/users.ts (hunk 1 above)
Comment (mia-pi-git) [reply server/users.ts:1483]: I'll make it `static` - That functionality _is_ needed in other places.

File: server/rooms.ts (hunk 2 above)
Comment (mia-pi-git) [reply server/rooms.ts]: Done.

File: server/rooms.ts (hunk 3 above)
Comment (mia-pi-git) [reply server/rooms.ts:1940]: Should this be? It seems fine where it is (also, passing the options down seems messy and not worth it?).

Comment (mia-pi-git) [reply server/rooms.ts]: Never mind, I just went and did it.

File: server/rooms.ts [hunk 4]
```diff
@@ -1922,39 +1922,11 @@ export const Rooms = {
...
-					room.settings.modjoin = '%';
-					room.add(`|raw|<div class="broadcast-red"><strong>This battle is invite-only!</strong><br />Users must be invited with <code>/invite</code> (or be staff) to join</div>`);
-				}
-			}
-		}
+		battle.checkPrivacySettings(options);
 
 		for (const p of players) {
 			if (p) {
+				battle.checkForcedUserSettings(p);
```
Comment (Zarel) [inline server/rooms.ts:1929]: This can also be moved inside `checkPrivacySettings`, right?

File: server/room-battle.ts [hunk 5]
```diff
@@ -16,6 +16,7 @@ import {execSync} from "child_process";
 import {BattleStream} from "../sim/battle-stream";
 import * as RoomGames from "./room-game";
 import type {Tournament} from './tournaments/index';
+import {RoomSettings} from './rooms';
```
Comment (Zarel) [inline server/room-battle.ts:19]:
```suggestion
import type {RoomSettings} from './rooms';
```

File: server/room-battle.ts [hunk 6]
```diff
@@ -1070,14 +1071,68 @@ export class RoomBattle extends RoomGames.RoomGame {
...
+		const privacySetter = new Set<ID>([]);
+		for (const p of ['p1', 'p2', 'p3', 'p4'] as const) {
+			const playerOptions = options[p];
+			if (playerOptions && playerOptions.inviteOnly) {
+				inviteOnly = true;
+				privacySetter.add(playerOptions.user.id);
+			} else if (playerOptions && playerOptions.hidden) {
+				privacySetter.add(playerOptions.user.id);
+			}
+			if (playerOptions?.user) this.checkForcedUserSettings(playerOptions.user);
```
Comment (Zarel) [inline server/room-battle.ts:1090]:
```suggestion
			if (playerOptions) {
				if (playerOptions.inviteOnly) {
					inviteOnly = true;
					privacySetter.add(playerOptions.user.id);
				} else if (playerOptions.hidden) {
					privacySetter.add(playerOptions.user.id);
				}
				if (playerOptions.user) this.checkForcedUserSettings(playerOptions.user);
			}
```

**Review by Zarel — APPROVED** [review]

**Comment by Zarel** [comment]
ty!

This will be merged for the restart tonight.

**Pull request merged**